grammar.js

```diff
@@ -67,6 +84,32 @@ module.exports = grammar({
         optional(";"),
       ),
 
+    
+    _ctes: $ => seq(
+      kw('WITH'),
+      optional(kw('RECURSIVE')),
+      commaSep1($.cte),
+    ),
+
+    cte: $ => seq(
+      $.identifier,
+      kw('AS'),
+      optional(seq(optional(kw('NOT')), kw('MATERIALIZED'))),
+      '(',
+      choice(
+        $.select_statement,
+        $.delete_statement,
+        $.insert_statement,
+        $.update_statement,
+      ),
+      ')',
+    ),
+
+    select_statement: $ => seq(optional($._ctes), $._select_statement),
```

Comment:
I think that instead statements having an optional `WITH` clause, they should instead be included to CTE:
```js
cte: $ => seq(
      $.identifier,
      kw('AS'),
      optional(seq(optional(kw('NOT')), kw('MATERIALIZED'))),
      '(',
      choice(
        $.select_statement,
        $.delete_statement,
        $.insert_statement,
        $.update_statement,
      ),
      ')',
      // optional?
      choice(
        $.select_statement,
        $.delete_statement,
        $.insert_statement,
        $.update_statement,
      ),
    ),
```

Reply:
According to the docs, in fact, the `SELECT`, `UPDATE`, `DELETE` and `INSERT` statements may have an optional `WITH` clause.

And the `WITH` clause may contains multiple `with_query`s (the `cte` node here).

Ref: https://www.postgresql.org/docs/14/sql-insert.html